extmod/mbedtls: Support Alternative Functions Implemented in Python.

iabdalkader opened the pull request:
### Summary

This patch enables the implementation of alternative mbedtls cryptography functions, such as ECDSA sign and verify, in pure Python. Alternative functions are implemented in Python callbacks, that get invoked from
wrapper functions when needed. If the callback fails to sign using the provided key, the calling code falls back to the default mbedtls function. A common use case for this feature is with secure elements that have drivers implemented in Python. Currently, only the ECDSA alternate sign function wrapper is implemented.

### Testing

I tested signing using a private EC key stored on a secure element (NXP's SE05x) and with the key stored on the filesystem (DER) both work fine. 

### Trade-offs and Alternatives

An alternative is adding the drivers/libraries of every secure element we wish to support, make them portable and provide port-specific code, as implemented in my previous PR: https://github.com/micropython/micropython/pull/15817

### Example usage:

```Python
import tls
import se05x

def ecdsa_sign_alt(key, data):
    if key[0:5] != b"magic":
        if log_level_enabled(logging.DEBUG):
            logging.debug("ecdsa_sign_alt falling back to default sign")
        raise RuntimeError("Not supported")

    objid = int.from_bytes(key[-4:], "big")
    if log_level_enabled(logging.DEBUG):
        logging.debug(f"ecdsa_sign_alt oid: 0x{objid:02X}")

    # Sign data on SE using reference key object id.
    sig = se_dev.sign(objid, data)
    if log_level_enabled(logging.DEBUG):
        sig_hex = "".join("%02X" % b for b in sig)
        logging.debug(f"ecdsa_sign_alt sig: {sig_hex}")
    logging.info("signed using secure element")
    return sig

# Create and initialize SE05x device.
se_dev = se05x.SE05X()

# Create SSL context etc...
ctx = tls.SSLContext(ssl.PROTOCOL_TLS_CLIENT)

# Set alternate ECDSA sign function.
ctx.ecdsa_sign_alt = ecdsa_sign_alt

# etc...
```

iabdalkader committed:
extmod/modtls_mbedtls: Support alternate sign callbacks in Python.

This commit enables the implementation of alternative mbedTLS cryptography
functions, such as ECDSA sign and verify, in pure Python.  Alternative
functions are implemented in Python callbacks, that get invoked from
wrapper functions when needed.  The callback can return None to fall back
to the default mbedTLS function.

A common use case for this feature is with secure elements that have
drivers implemented in Python.  Currently, only the ECDSA alternate sign
function wrapper is implemented.

Tested signing with a private EC key stored on an NXP SE05x secure element.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32/boards/ARDUINO_PORTENTA_H7: Add SE05x driver.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32/boards/ARDUINO_NICLA_VISION: Add SE05x driver.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented on extmod/mbedtls/mbedtls_alt.c:
```diff
@@ -0,0 +1,84 @@
+
+    // Try signing with the alternate function first, then fallback to the default.
+    if (mbedtls_ecdsa_sign_alt(&ctx->grp, &r, &s, &ctx->d, hash, hlen, f_rng, p_rng) != 0) {
+        MBEDTLS_MPI_CHK(mbedtls_ecdsa_sign(&ctx->grp, &r, &s, &ctx->d, hash, hlen, f_rng, p_rng));
```
Note to avoid duplicating a lot of code `MBEDTLS_ECDSA_SIGN_ALT` is NOT defined to keep the default `mbedtls_ecdsa_sign` to use as a fallback function. However, `mbedtls_ecdsa_sign` can't be wrapped, because it's called internally to its object file, so we wrap `mbedtls_ecdsa_write_signature` and `mbedtls_ecdsa_read_signature` instead.

projectgus replied:
It's annoying there's no way to keep the default implementation without a linker wrap. I just went down the rabbit hole of looking at replacing the `mbedtls_eckey_info` structure instead (as it's the only place inside mbedTLS which calls through to here), but I can't see a different way there either.

Could you please put some of your rationale from the comment here into the top-level comment for `__wrap_mbedtls_ecdsa_write_signature`? That way it'll be clearer for whoever touches this next.

iabdalkader replied:
@projectgus I also tried many different things, this is the only way right now. And yes it's extremely annoying, I don't know why they didn't just make the functions weak like everyone else and saved us the trouble.

iabdalkader replied:
Done.

iabdalkader commented on extmod/mbedtls/mbedtls_alt.c:
```diff
@@ -0,0 +1,84 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
This file somewhat follows code from `ecdsa.c` which is Apache/GPL.  Let me know if the license is okay.

iabdalkader replied:
This file could contain multiple other wrappers. I've added a line just above the wrapper function if it helps:
// See lib/mbedtls/library/ecdsa.c:688

iabdalkader committed:
lib/arduino-lib: Update submodule.

Changes:
- Add ISO7816, APDU and SE05x package.
- Add support for Opta Expansion protocol.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented on extmod/modtls_mbedtls.c:
```diff
@@ -767,6 +771,76 @@ static MP_DEFINE_CONST_OBJ_TYPE(
 // ssl module.
 
+#if MICROPY_PY_SSL_ECDSA_SIGN_ALT
+static mp_obj_t ecdsa_sign_alt_cb = mp_const_none;
```
```suggestion
static mp_obj_t ecdsa_sign_alt_cb = mp_const_none;
```
This doesn't need to be a root pointer, right ?

dpgeorge replied:
Yes it will need to be a root pointer.  Eg:
```py
tls.ecdsa_sign_alt(lambda k, b: ...)
```
In that case there is no other reference to the lambda object other than this pointer here, so it must be a root pointer.

iabdalkader replied:
Good point! It's unlikely to be passed as a lambda, but I made it a root pointer anyway.

iabdalkader commented:
Anything else needs to change here ?

dpgeorge commented:
I agree that this is more general than #15817.

Did you consider putting the new function on the `SSLContext` object?  Eg `SSLContext.ecdsa_sign_callback`?  That way, each SSLContext could have a different alt implementation, which would be even more flexible.  Eg some connections could use the default, and others could use the custom one.  (I'm not sure exactly how this would be implemented, because there needs to be a pointer to the correct SSLContext, but it's worth thinking about this possibility.)

iabdalkader commented on extmod/modtls_mbedtls.c:
```diff
@@ -767,6 +771,74 @@ static MP_DEFINE_CONST_OBJ_TYPE(
+    uint8_t key[256];
+
+    if (MP_STATE_PORT(ecdsa_sign_alt_cb) == mp_const_none) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
```
The fallback is to allow signing with a key that's Not stored on the SE (a DER on the filesystem for example). The SE can't be mutually exclusive, you may not want to use it all even if it's enabled.

The alternative is to either check if the key has magic bytes before calling the SE, or to return a special value to indicate that this is a regular key, not stored in SE.

dpgeorge replied:
OK, so that's a third use case:
- alternate function provided but it decides to fall back to the default


> or to return a special value to indicate that this is a regular key, not stored in SE.

Yes, I think it should do that.  So you explicitly indicate that you want to fall back to the default.  I suggest returning `None` from the Python callback to indicate this.  Any exception that's raised by the Python callback should be reported to the user and abort the function.

iabdalkader replied:
Done. If the callback returns `None` the wrapper will fall back to the default implementation. If it raises an exception, the exception is caught, printed and the proxy function returns an error to the wrapper.

iabdalkader commented:
> Did you consider putting the new function on the `SSLContext` object? Eg `SSLContext.ecdsa_sign_callback`?
> I'm not sure exactly how this would be implemented, because there needs to be a pointer to the correct SSLContext, but it's worth thinking about this possibility

Yes I did, that's actually how I wanted to implement this originally, but quickly realized that there's no way to get the SSL object in mbedtls functions. It's not an issue though, because that's what the fallback is for; if another context uses regular keys, it will still function.

FWIW this is how this feature is currently used:

https://github.com/arduino/arduino-iot-cloud-py/blob/new_se_support/src/arduino_iot_cloud/ussl.py#L34

This is not merged, waiting on it to be merged upstream first.

dpgeorge commented:
> Yes I did, that's actually how I wanted to implement this originally, but quickly realized that there's no way to get the SSL object in mbedtls functions.

I think we should try to implement it this way, as an attribute per `SSLContext`.  It's not that difficult actually, just have a global (thread local in threaded systems) variable which points to the current `SSLContext` object and be sure to set that variable on entry to all C functions in `modtls_mbedtls.c` that will call into mbedTLS code (eg `socket_read` and `socket_write` are the main ones).  Then when you need to retrieve the signing callback, access this global `SSLContext` variable.

I think that's a lot more flexible, and would interoperate well with other extensions like DTLS #15764 which may not want to use custom signing.

dpgeorge commented:
> I'll get around to making all of the requested changes next week

OK, great, thanks!

I think having the callback on the `SSLContext` will be great if you can get that working.  Here are some suggestions for that:
```c
// in py/mpconfig.h
#define MICROPY_PY_SSL_MBEDTLS_NEED_ACTIVE_CONTEXT (MICROPY_PY_SSL_ECDSA_SIGN_CALLBACK)

// in py/mpstate.h, mp_state_thread_t struct
    #if MICROPY_PY_SSL_MBEDTLS_NEED_ACTIVE_CONTEXT
    struct _mp_obj_ssl_context_t *tls_ssl_context;
    #endif

// in extmod/modtls_mbedtls.c
static inline void store_active_context(mp_obj_ssl_context_t *ssl_context) {
    #if MICROPY_PY_SSL_MBEDTLS_NEED_ACTIVE_CONTEXT
    MP_THREAD_STATE(tls_ssl_context) = ssl_context;
    #endif
}

// at start of functions that call into mbedtls
    store_active_context(ssl_context);

// could also do this at end of functions, but not necessary
    store_active_context(NULL);
```

iabdalkader committed:
stm32/boards/ARDUINO_OPTA: Add Opta expansion module.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented:
@dpgeorge Can we add the callback attribute to `ssl.py` as well ? It's non-standard, but would make it much more convenient to set from code that uses `ssl` for portability. Alternatively, it could probably be set using something like:
```
ctx._context.ecdsa_sign_callback = callback
```

iabdalkader commented on extmod/extmod.mk:
```diff
@@ -242,6 +243,10 @@ MBEDTLS_CONFIG_FILE ?= \"mbedtls/mbedtls_config_port.h\"
 GIT_SUBMODULES += $(MBEDTLS_DIR)
 CFLAGS_EXTMOD += -DMBEDTLS_CONFIG_FILE=$(MBEDTLS_CONFIG_FILE)
 CFLAGS_EXTMOD += -DMICROPY_SSL_MBEDTLS=1 -I$(TOP)/$(MBEDTLS_DIR)/include
+ifeq ($(MICROPY_PY_SSL_ECDSA_SIGN_ALT),1)
```
I think the current name is better as it reflects that it enables a MicroPython-specific alt sign function, not a callback for it or for the the standard sign.

dpgeorge replied:
OK

iabdalkader committed:
extmod/modtls_mbedtls: Add a thread-global ptr for current SSL context.

This is necessary for mbedTLS callbacks that do not carry any user state,
so those callbacks can be customised per SSL context.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented on ports/stm32/boards/ARDUINO_NICLA_VISION/mbedtls_config_board.h:
```diff
@@ -2,6 +2,7 @@
 #define MICROPY_INCLUDED_MBEDTLS_CONFIG_BOARD_H
 
 #define MBEDTLS_ECP_NIST_OPTIM
+#define MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED
```
Removed.

dpgeorge commented:
> Can we add the callback attribute to `ssl.py` as well ? It's non-standard, but would make it much more convenient to set from code that uses `ssl` for portability.

We are trying hard to keep `ssl` compatible with CPython, so we should not add this callback there.

I'm not sure what you mean by "portability" here?  It won't be portable to CPython, and any existing MicroPython code needs to update to firmware that supports this callback, and that firmware will have the new `tls` module.

dpgeorge commented:
> If the root pointer is not cleared, would this stop the socket from being collected after its closed ?

It will stop the `SSLContext` from being collected, but I think the socket will still be freed because it won't have a reference (I don't think).

iabdalkader commented:
> are trying hard to keep `ssl` compatible with CPython, so we should not add this callback there.

Ah I sent a PR but will remove it.

> I'm not sure what you mean by "portability" here? It won't be portable to CPython, and any existing MicroPython code needs to update to firmware that supports this callback, and that firmware will have the new `tls` module

I usually just import the `ssl` module on CPython and MicroPython, not `tls`. To access this attribute from `tls` I would have to switch to it or use `ctx._context.ecdsa_sign_callback = callback` ? The latter is fine.

> It will stop the SSLContext from being collected

I've updated the code, it now stores the context before certain functions, read, write, wrap_socket, and clears it on stream close. Will that be enough ?

iabdalkader commented on extmod/modtls_mbedtls.c:
```diff
@@ -767,6 +771,74 @@ static MP_DEFINE_CONST_OBJ_TYPE(
+    }
+
+    // Convert the MPI private key (d) to a binary array
+    ret = mbedtls_mpi_write_binary(d, key, klen);
```
I'm not sure, but I think it's more efficient this way, note the bytearrays are passed by reference.

iabdalkader commented on py/mpconfig.h:
```diff
@@ -1814,6 +1814,11 @@ typedef double mp_float_t;
 #define MICROPY_PY_SSL_FINALISER (MICROPY_ENABLE_FINALISER)
 #endif
 
+// Whether to add a root pointer for the current ssl object
+#ifndef MICROPY_PY_SSL_MBEDTLS_NEED_ACTIVE_CONTEXT
+#define MICROPY_PY_SSL_MBEDTLS_NEED_ACTIVE_CONTEXT (MICROPY_PY_SSL_ECDSA_SIGN_ALT)
+#endif
+
```
Can't this just be a REGISTER_ROOT_POINTER in modtls_mbedtls.c ?

dpgeorge replied:
No, there needs to be a separate variable per thread.  Because each thread may be using a different `SSLContext` with a different callback.

iabdalkader replied:
Ah I see.

iabdalkader commented on extmod/modtls_mbedtls.c:
```diff
@@ -767,6 +771,74 @@ static MP_DEFINE_CONST_OBJ_TYPE(
+        mp_get_buffer_raise(ret, &sig, MP_BUFFER_READ);
+        nlr_pop();
+    } else {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
```
Done.

iabdalkader commented on extmod/mbedtls/mbedtls_alt.c:
```diff
@@ -0,0 +1,95 @@
+    size_t len = 0;
+    MBEDTLS_ASN1_CHK_ADD(len, mbedtls_asn1_write_mpi(&p, buf, &s));
+    MBEDTLS_ASN1_CHK_ADD(len, mbedtls_asn1_write_mpi(&p, buf, &r));
+
```
Yes, the `return 0;` is extra, it should fall through to the cleanup label.

iabdalkader commented on extmod/mbedtls/mbedtls_alt.c:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    size_t len = 0;
+    MBEDTLS_ASN1_CHK_ADD(len, mbedtls_asn1_write_mpi(&p, buf, &s));
```
> This probably needs to be MBEDTLS_ASN1_CHK_CLEANUP_ADD?

Yes this makes sense because in this function there's a `cleanup` label.

> Also, could this function copy directly into sig,

Here I'm just following the reference implementation, see `ecdsa_signature_to_asn1`. I'm a bit reluctant to change this code. It's not any less efficient than the default implementation and the buffer is stack based so no dynamic memory.

iabdalkader commented on extmod/modtls_mbedtls.c:
```diff
@@ -767,6 +797,76 @@ static MP_DEFINE_CONST_OBJ_TYPE(
+    uint8_t *p = (uint8_t *)sig.buf;
+    uint8_t *e = (uint8_t *)sig.buf + sig.len;
+
+    // Parse the DER-encoded signature to extract r and s
```
> Can we just skip the whole decoding/encoding and just return the ASN.1 data directly from this callback

We can, I guess even if a device doesn't output ASN.1 directly, it could still encode it in Python.

iabdalkader replied:
> Can we just skip the whole decoding/encoding and just return the ASN.1

I now return the ASN.1 data directly without decoding/re-encoding.

iabdalkader commented on extmod/modtls_mbedtls.c:
```diff
@@ -68,6 +72,7 @@ typedef struct _mp_obj_ssl_context_t {
     int authmode;
     int *ciphersuites;
     mp_obj_t handler;
+    mp_obj_t ecdsa_sign_callback;
```
Done.

dpgeorge commented:
This will eventually need docs... but we somehow forgot to document the `tls` module when that was added, and doing that is out of scope for this PR.  So docs for this alternative signing callback can be added when `tls` itself is documented.

Also would be good to add some tests for this new functionality, that can be run with the unix coverage build.  That's also very difficult, it requires at least having a loopback socket and creating a dummy TLS connection.  For now, testing on hardware is good enough.
